File: Source/WinFormsApp/PseudoRandomSignal.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace WindowsFormsApp
{
    public class PseudoRandomSignal
    {
        public double[] FrequencyList { get; private set; }
        public int NSequence { get; private set; }
        private double[] Sequence { get; set; } = Array.Empty<double>();

        public PseudoRandomSignal(double[] frequencyList, string method = "default")
        {
            if (frequencyList == null || frequencyList.Length == 0)
            {
                frequencyList = new double[] { 1 };
            }

            foreach (var freq in frequencyList)
            {
                if (freq <= 0 || freq != Math.Floor(freq))
                {
                    throw new ArgumentException("frequencyList must contain positive integers.");
                }
            }

            switch (method)
            {
                case "2n":
                    if (frequencyList.Length != 1)
                    {
                        throw new ArgumentException("For '2n' method, frequencyList should be a scalar.");
                    }
                    int n = (int)frequencyList[0];
                    FrequencyList = new double[n];
                    for (int i = 0; i < n; i++)
                    {
                        FrequencyList[i] = Math.Pow(2, i);
                    }
                    MakeSequence();
                    break;
                case "pattern":
                    Sequence = new double[frequencyList.Length * 2];
                    for (int i = 0; i < frequencyList.Length; i++)
                    {
                        Sequence[i] = frequencyList[i];
                        Sequence[i + frequencyList.Length] = -frequencyList[i];
                    }
                    NSequence = Sequence.Length;
                    FrequencyList = new double[] { 1 };
                    break;
                default:
                    FrequencyList = frequencyList;
                    MakeSequence();
                    break;
            }
        }
        
        public double[] Sampling(int nInterpolation = 1, int samplingTime = 1)
        {
            if (nInterpolation <= 0 || samplingTime <= 0)
            {
                throw new ArgumentException("nInterpolation and samplingTime must be positive integers.");
            }

            var repeatedSequence = new List<double>();
            foreach (var item in Sequence)
            {
                for (int i = 0; i < nInterpolation; i++)
                {
                    repeatedSequence.Add(item);
                }
            }

            var result = new List<double>();
            for (int i = 0; i < samplingTime; i++)
            {
                result.AddRange(repeatedSequence);
            }

            return result.ToArray();
        }

        private void MakeSequence()
        {
            int nFrequency = FrequencyList.Length;
            int nColsHalfMatrix = (int)FrequencyList[0];

            for (int i = 1; i < nFrequency; i++)
            {
                nColsHalfMatrix = LCM(nColsHalfMatrix, (int)FrequencyList[i]);
            }

            int nColsMatrix = nColsHalfMatrix * 2;
            double[,] sequenceMatrix = new double[nFrequency, nColsMatrix];
            Sequence = new double[nColsMatrix];
            NSequence = nColsMatrix;

            for (int i = 0; i < nFrequency; i++)
            {
                int freq = (int)FrequencyList[i];
                int halfPeriodLength = nColsHalfMatrix / freq;

                double[] halfPeriod = new double[halfPeriodLength];
                for (int j = 0; j < halfPeriodLength; j++)
                {
                    halfPeriod[j] = 1;
                }

                double[] pattern = new double[nColsHalfMatrix];
                Array.Copy(halfPeriod, 0, pattern, 0, halfPeriodLength);
                for (int j = 0; j < halfPeriodLength; j++)
                {
                    pattern[j + halfPeriodLength] = -1;
                }

                double[] row = new double[nColsMatrix];
                for (int j = 0; j < freq; j++)
                {
                    Array.Copy(pattern, 0, row, j * nColsHalfMatrix, nColsHalfMatrix);
                }

                for (int j = 0; j < nColsMatrix; j++)
                {
                    sequenceMatrix[i, j] = row[j];
                }
            }

            for (int j = 0; j < nColsMatrix; j++)
            {
                double sum = 0;
                for (int i = 0; i < nFrequency; i++)
                {
                    sum += sequenceMatrix[i, j];
                }
                Sequence[j] = Math.Sign(sum);
            }
        }

        private static int LCM(int a, int b)
        {
            return a / GCD(a, b) * b;
        }

        private static int GCD(int a, int b)
        {
            while (b != 0)
            {
                int temp = b;
                b = a % b;
                a = temp;
            }
            return a;
        }
    }
}

File: Source/cmdApp/method/PseudoRandomSignal.m
classdef PseudoRandomSignal < handle
    % PseudoRandomSignal 类（句柄类）
    % 用于生成一种基于多个频率成分的伪随机二值（±1）信号序列

    properties (GetAccess = public, SetAccess = private)
        frequencyList % 输入的频率成分列表，例如 [1, 2, 4]
        nSequence     % 生成的信号序列的总长度（矩阵列数）
    end

    properties (Access = private)
        sequence % 最终生成的伪随机信号序列，元素为 ±1
    end

    methods (Access = public)
        % ===== 构造函数 =====
        function prs = PseudoRandomSignal(frequencyList)
            arguments
                frequencyList (1, :) double = 1  % 默认频率为 1 Hz
            end
            prs.frequencyList = frequencyList;  % 保存频率列表
            prs.MakeSequence();  % 调用方法生成信号序列
        end

        % ===== 重采样方法（生成更长的信号）=====
        function signal = Sampling(prs, n_interpolation, sampling_time)
            arguments
                prs             (1,1) PseudoRandomSignal
                n_interpolation (1,1) double = 1  % 插值倍数
                sampling_time   (1,1) double = 1  % 采样时间倍数
            end

            
            % 通过重复原始序列的元素，生成更长信号
            signal = repmat(repelem(prs.sequence, n_interpolation), ...
                                    1, sampling_time);
        end
    end






    methods (Access = private)
        % ===== 核心：生成伪随机信号序列 =====
        function prs = MakeSequence(prs)
            nFrequency         = length(prs.frequencyList);  % 频率个数
            n_cols_half_matrix = prs.frequencyList(1);       % 初始周期设为第一个频率的周期数
            for i = 2:nFrequency
                % 计算所有频率的最小公倍数，作为基本周期长度
                n_cols_half_matrix = lcm(n_cols_half_matrix, prs.frequencyList(i));
            end
            n_cols_matrix  = n_cols_half_matrix*2;  % 总矩阵宽度（双倍周期？）
            sequenceMatrix = zeros(nFrequency, n_cols_matrix);  % 初始化频率×时间的矩阵
            prs.sequence   = zeros(1, n_cols_matrix);  % 最终信号序列
            prs.nSequence  = n_cols_matrix;  % 信号长度

            % 对每个频率，构造一个周期内的方波模式
            for i = 1:nFrequency
                half_period          = ones(1, n_cols_half_matrix/prs.frequencyList(i));
                pattern              = cat(2, half_period, -half_period);  % 前半为1，后半为-1
                sequenceMatrix(i, :) = repmat(pattern, 1, prs.frequencyList(i));  % 按频率重复
            end

            % 将所有频率的波形叠加后取符号，生成最终 ±1 序列
            prs.sequence = sign(sum(sequenceMatrix, 1));
        end
    end
end